### ISSUE TITLE
`pub` in executable modules

### ISSUE THREAD

**Issue by BatmanAoD** [rust-lang/rust-clippy#13644]
### What it does

Warn of uses of `pub` in the top-level module of an executable-only target.

Specifically, `pub` outside of a `mod` block in any of these files would trigger the lint:

* `src/main.rs`
* `src/bin/foo.rs`
* `src/bin/foo/main.rs`

### Advantage

- `pub` suppresses "unused" lints, so inadvertently marking something `pub` may mask that the user forgot to use it.
- `pub` has no other effect (that I'm aware of) in top-level modules of executable-only targets.

I actually ran into this recently: I work on a project that has several executable targets, and I discovered that two of them were not actually using all of their declared arguments.

### Drawbacks

There may be a use for `pub` inside of an executable-only module that I'm not aware of.

### Example

```rust
use clap::Parser;

#[derive(Parser, Debug)]
struct Args {
    #[arg(short, long)]
    pub name: String,
}
```

Could be written as:

```rust
use clap::Parser;

#[derive(Parser, Debug)]
struct Args {
    #[arg(short, long)]
    name: String,
}
```

**Comment by danya02** [comment]
I had a similar issue, where I wrote Bevy system functions as `pub fn` and then forgot to register them in schedules, meaning that a new feature I wrote didn't work and I didn't know why.

I ended up rewriting all `pub` to `pub(crate)` in my application crate, but I'd like a lint that would either tell me to do this, or (better) to warn about unused `pub` items regardless.
